리뷰 작성 수에 따른 회원등급 변경 #41

nahokyun opened the pull request:
> ### 변경된점
- Grade 변경요청이 들어오면 사용자의 작성 review수를 체크한뒤 5개가 넘으면 VIP, 5개 이하라면 BASIC으로 변경하는 기능이 추가되었습니다.
## MovieController.java
### 추가된 부분
```java
    @PostMapping("/change-grade")
    @ResponseStatus(HttpStatus.OK)
    @LoginCheck
    public String changeGrade(){
        reviewService.checkGrade();
        return SUCCESS_CHANGE_GRADE.getMessage();
    }

```
## ReviewService.java
### 추가된 함수 
```java
    public void checkGrade(){
        Long memberId = (Long) httpSession.getAttribute(SessionConst.MEMBER_ID);

        int reviewCount=reviewMapper.getReviewCount(memberId);

        if(reviewCount>5)
            memberMapper.changeGrade(memberId, Grade.VIP);
        else
            memberMapper.changeGrade(memberId,Grade.BASIC);
    }

```
## ReviewMapper.xml
### 추가된 쿼리
```xml
    <select id="getReviewCount" resultType="int">
        SELECT COUNT(*)
        FROM review
        WHERE member_id = #{memberId}
    </select>

```
## MemberMapper.xml
### 추가된 쿼리
```xml
    <update id="changeGrade" >
        update member
        set grade=#{grade}
        where member.id = #{memberId}
    </update>
```

nahokyun committed:
리뷰 작성 수에 따른 회원등급 변경 #41

Grade 변경요청이 들어오면 사용자의 작성 review수를 체크한뒤 5개가 넘으면 VIP, 5개 이하라면 BASIC으로 변경하는 기능이 추가되었습니다.

sunwootest commented on src/main/java/study/movieservice/controller/MovieController.java:
```diff
@@ -77,4 +77,12 @@ public String recommendDelete(@RequestBody Recommend recommend) {
         return SUCCESS_RECOMMEND_DELETE.getMessage();
     }
+
+    @PostMapping("/change-grade")
+    @ResponseStatus(HttpStatus.OK)
+    @LoginCheck
+    public String changeGrade(){
+        reviewService.checkGrade();
+        return SUCCESS_CHANGE_GRADE.getMessage();
+    }
```
변경에 대한 API는 POST 방식보단 다른 방식이 더 나아보이네요 :) 굳이 change를 붙일 필요도 없어지고요.

nahokyun replied:
PatchMapping을 사용하여 변경을 나타내는 쪽으로 수정하는것이 더 나을것 같아 반영하겠습니다!

sunwootest commented on src/main/java/study/movieservice/service/ReviewService.java:
```diff
@@ -76,4 +82,14 @@ public void decreaseDislikeCount(Long reviewId){
     public Long getLikeCount(Long reviewId){
         return reviewMapper.getLikeCount(reviewId);
     }
+    public void checkGrade(){
+        Long memberId = (Long) httpSession.getAttribute(SessionConst.MEMBER_ID);
+
+        int reviewCount=reviewMapper.getReviewCount(memberId);
+
+        if(reviewCount>5)
+            memberMapper.changeGrade(memberId, Grade.VIP);
+        else
+            memberMapper.changeGrade(memberId,Grade.BASIC);
+    }
```
메소드 명은 Grade를 확인하는 것 같은데 Grade를 변경하기도 하네요. 이 경우에는 트랜잭션이 필요하지 않을까요? 컨벤션도 지켜주세요.

sunwootest commented on src/main/resources/study/movieservice/repository/ReviewMapper.xml:
```diff
@@ -74,5 +74,11 @@
         where movie_id = #{movieId}
     </select>
+
+    <select id="getReviewCount" resultType="int">
+        SELECT COUNT(*)
+        FROM review
+        WHERE member_id = #{memberId}
+    </select>
```
soft delete를 적용하지 않을 것인지 모르겠지만 나중에 추가하려면 귀찮아집니다.

nahokyun replied:
review와 recommend 테이블의 삭제기능을 hard delete에서 soft delete로 변경하는것을 다음 이슈로 처리하도록 하겠습니다!

nahokyun committed:
리뷰 작성 수에 따른 회원등급 변경 #41

checkGrade Transactional을 적용하였습니다.

changeGrade의 Rest Api방식을 Post에서 Patch로 변경하였습니다.

nahokyun commented:
> ### 변경된점
- RestApi의 매핑 방식을 PostMapping에서 PatchMapping으로 변경하였습니다.
- checkGrade에 Transaction을 적용하였습니다.
## MovieController.java
### 변경 전 
```java
    @PostMapping("/change-grade")
    @ResponseStatus(HttpStatus.OK)
    @LoginCheck
    public String changeGrade(){
        reviewService.checkGrade();
        return SUCCESS_CHANGE_GRADE.getMessage();
    }
```
### 변경 후 
```java
    @PatchMapping("/grade")
    @ResponseStatus(HttpStatus.OK)
    @LoginCheck
    public String changeGrade(){
        reviewService.checkGrade();
        return SUCCESS_CHANGE_GRADE.getMessage();
    }

```
## ReviewService.java
### 변경 전 
```java
    public void checkGrade(){
        Long memberId = (Long) httpSession.getAttribute(SessionConst.MEMBER_ID);

```
### 변경 후 
```java
    @Transactional
    public void checkGrade(){

        Long memberId = (Long) httpSession.getAttribute(SessionConst.MEMBER_ID);
  
```
